plugins/sharding/logical_count.rb

```diff
@@ -56,179 +41,134 @@ def cache_key(input)
+        def execute
+          ensure_filtered
+
+          if @range_index
+            return count_n_records_in_range
           end
-          message << " <#{table_name}>: #{reason}"
-          logger.log(Logger::Level::DEBUG,
-                     __FILE__,
-                     line,
-                     method.to_s,
-                     message)
-        end
 
-        def prepare_contexts
-          if @filter or @post_filter
-            if @dynamic_columns.have_initial?
-              apply_targets = []
-              @contexts.each do |context|
-                table = context.table.select_all
-                @temporary_tables << table
-                context.table = table
-                apply_targets << [table]
+          return 0 if @filtered_result_sets.empty?
+
+          total = 0
+          if @window
+            @filtered_result_sets.each do |result_set|
+              @window.each(result_set) do |windowed_result_set|
+                @temporary_tables << windowed_result_set
+                if @context.dynamic_columns.have_filtered?
+                  apply_targets = [[windowed_result_set]]
+                  @context.dynamic_columns.apply_filtered(apply_targets)
+                end
+                unless @post_filter.nil?
+                  windowed_result_set = apply_post_filter(windowed_result_set)
+                  @temporary_tables << windowed_result_set
+                end
+                total += windowed_result_set.size
               end
-              @dynamic_columns.apply_initial(apply_targets)
             end
-          end
-          @contexts.each do |context|
-            filter_shard(context)
-          end
-          if @post_filter
-            if @dynamic_columns.have_filtered?
-              apply_targets = @contexts.collect do |context|
-                [context.table]
+          else
+            apply_filtered_dynamic_columns
+            @filtered_result_sets.each do |result_set|
+              unless @post_filter.nil?
+                result_set = apply_post_filter(result_set)
+                @temporary_tables << result_set
               end
-              @dynamic_columns.apply_filtered(apply_targets)
-            end
-            @contexts.each do |context|
-              post_filter_shard(context)
+              total += result_set.size
             end
           end
+          return total
         end
 
-        def filter_shard(context)
-          return if context.range_index
+        def use_range_index?
+          if @filter or @post_filter
+            decide_use_range_index(false, "need filter",
+                                __LINE__, __method__)
+          elsif @cover_type == :all
+            decide_use_range_index(false, "covered",
+                                __LINE__, __method__)
+          else
+            decide_use_range_index(true, "range index is available",
+                                __LINE__, __method__)
+          end
```

Comment:
Currently the debug log `no range index` is not outputted.
I will fix it in another PR because we need to modify StreamShardExecutor (https://github.com/groonga/groonga/blob/master/plugins/sharding/stream_shard_executor.rb#L74) for fixing it and it affects to not only `logical_count` but also `logical_range_filter`.

Reply:
This is resolved by #1349

Reply:
This is resolved by #1349